compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala

```diff
@@ -61,44 +61,117 @@ object PickledQuotes {
+    val Inlined(call, Nil, expnasion) = unpickled
+    val inlineCtx = inlineContext(call)
+    val expansion1 = spliceTypes(expnasion, splices)(using inlineCtx)
+    val expansion2 = spliceTerms(expansion1, splices)(using inlineCtx)
+    cpy.Inlined(unpickled)(call, Nil, expansion2)
   }
 
   /** Unpickle the tree contained in the TastyType */
   def unpickleType(tasty: PickledQuote, args: PickledArgs)(implicit ctx: Context): Tree = {
     val tastyBytes = TastyString.unpickle(tasty)
     val unpickled = unpickle(tastyBytes, args, isType = true)(ctx.addMode(Mode.ReadPositions))
-    val tpt = unpickled match {
-      case Block(aliases, tpt) =>
-        // `@quoteTypeTag type` aliases are not required after unpickling.
-        // Type definitions are placeholders for type holes in the pickled quote, at this point
-        // those holes have been filled. As we already dealias al references to them in `dealiasTypeTags`
-        // there is no need to keep their definitions in the tree. As artifacts of quote reification
-        // they also do not have a meaningful position in the source.
-        val aliases1 = aliases.filter(!_.symbol.hasAnnotation(defn.InternalQuoted_QuoteTypeTagAnnot))
-        seq(aliases1, tpt)
-      case tpt => tpt
+    spliceTypes(unpickled, args)
+  }
+
+  /** Replace all term holes with the spliced terms */
+  private def spliceTerms(tree: Tree, splices: PickledArgs)(using Context): Tree = {
+    val evaluateHoles = new TreeMap {
+      override def transform(tree: tpd.Tree)(implicit ctx: Context): tpd.Tree = tree match {
+        case Hole(isTerm, idx, args) =>
+          val reifiedArgs = args.map { arg =>
+            if (arg.isTerm) (qctx: scala.quoted.QuoteContext) ?=> new TastyTreeExpr(arg, QuoteContext.scopeId)
+            else new TreeType(arg, QuoteContext.scopeId)
+          }
+          if isTerm then
+            val splice1 = splices(idx).asInstanceOf[Seq[Any] => scala.quoted.QuoteContext ?=> quoted.Expr[?]]
+            val quotedExpr = splice1(reifiedArgs)(using dotty.tools.dotc.quoted.QuoteContext())
+            val filled = PickledQuotes.quotedExprToTree(quotedExpr)
+
+            // We need to make sure a hole is created with the source file of the surrounding context, even if
+            // it filled with contents a different source file.
+            if filled.source == ctx.source then filled
+            else filled.cloneIn(ctx.source).withSpan(tree.span)
+          else
+            // Replaces type holes generated by ReifyQuotes (non-spliced types).
+            // These are types defined in a quote and used at the same level in a nested quote.
+            val quotedType = splices(idx).asInstanceOf[Seq[Any] => quoted.Type[?]](reifiedArgs)
+            PickledQuotes.quotedTypeToTree(quotedType)
+        case tree: Select =>
+          // Retain selected members
+          val qual = transform(tree.qualifier)
+          qual.select(tree.symbol).withSpan(tree.span)
```

Comment:
It would be nice to doc & spec this invariant: quoting & splicing do not change overloading resolution. The following can be an example to demonstrate:

```Scala
class A[T] {
  def foo(x: T): T = ???
  def foo(x: Int): Int = ???
}

class B extends A[Int]

import scala.quoted._

inline def m[T](a: A[T], x: T) : T = ${  mImpl[T]('a, 'x) }

def mImpl[T: Type](a: Expr[A[T]], x: Expr[T])(using qctx: QuoteContext): Expr[T] =  '{
  $a.foo($x)
}

// which `foo` will be called?
m(new B,  5)
```

Reply:
That is in part of the spec of inlining. We just did something else by mistake. I will try to place this in the docs.